### PR TITLE
Adding support for running tests on windows

### PR DESCRIPTION
**Pull request by vijetmahabaleshwar-okta**

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -71,6 +71,54 @@ $ git rebase master
...
+* [A SPA application](https://github.com/okta/samples-js-angular/tree/master/okta-hosted-login#prerequisites)
+* A test user account with a known username and password. Note that the USERNAME should be of the form "username@email.com"
+
+Once you have those resources setup, export their details as the following environment variables:
+
+```bash
+export ISSUER=https://{yourOktaDomain}.com/oauth2/default
+export CLIENT_ID={yourWebAppClientId}
+export CLIENT_SECRET={yourWebAppClientSecret}
+export SPA_CLIENT_ID={yourSpaAppClientId}
```
Comment (bretterer) [inline CONTRIBUTING.md:96]: Do we actually need a spa client id?  I was able to do all of this with the same client_id / application as my web side app in testing.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -71,6 +71,54 @@ $ git rebase master
...
+export USERNAME={userName}
+export PASSWORD={password}
+```
+
+For Windows, please set the following environment variables:
+- `ISSUER`
+- `CLIENT_ID`
+- `CLIENT_SECRET`
+- `SPA_CLIENT_ID`
```
Comment (bretterer) [inline CONTRIBUTING.md]: see above

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -71,6 +71,54 @@ $ git rebase master
...
+- `USER_NAME`
+- `PASSWORD`
+
+> **NOTE:** Windows has USERNAME as a built-in system variable, hence set the USER_NAME environment variable for testing.
+
+Then run the E2E tests:
+
+```bash
+npm test
```
Comment (bretterer) [inline CONTRIBUTING.md]: all tests for the samples should be set to run from a `composer` command, not from npm itself.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -71,6 +71,54 @@ $ git rebase master
...
+```
+
+> **NOTE:** If you want to execute individual tests such as `npm run test:okta-hosted-login`, you will need to update the environment by running the following node script first:
+
+```bash
+node scripts/setup-env.js
```
Comment (bretterer) [inline CONTRIBUTING.md]: Can all this be done directly in the composer.json file instead of having to run more things?   

composer scripts

```
  'test:okta-hosted-login': "node scripts/setup-env.js && ......"
```

File: composer.json [hunk 5]
```diff
@@ -27,28 +27,27 @@
             "echo \"Updating Resource Server Dependencies\" && cd resource-server && composer update"
         ],
         "test": [
-            "composer update-okta-oidc-tck",
```
Comment (bretterer) [inline composer.json:30]: Why did you remove the update-okta-oidc-tck?

File: composer.json [hunk 6]
```diff
@@ -27,28 +27,27 @@
...
+            "composer setup-env",
             "composer test:custom-login",
             "composer test:okta-hosted-login",
             "composer test:resource-server"
         ],
         "test:custom-login": [
-            "composer update-okta-oidc-tck",
```
Comment (bretterer) [inline composer.json]: Why did you remove the update-okta-oidc-tck?

File: composer.json [hunk 7]
```diff
@@ -27,28 +27,27 @@
...
+            "composer setup-env",
             "if [[ -f testenv ]]; then source testenv; fi && npm run test:custom-login"
         ],
         "test:okta-hosted-login": [
-            "composer update-okta-oidc-tck",
```
Comment (bretterer) [inline composer.json]: Why did you remove the update-okta-oidc-tck?

File: composer.json [hunk 8]
```diff
@@ -27,28 +27,27 @@
...
+            "composer setup-env",
             "if [[ -f testenv ]]; then source testenv; fi && npm run test:okta-hosted-login"
         ],
         "test:resource-server": [
-            "composer update-okta-oidc-tck",
```
Comment (bretterer) [inline composer.json]: Why did you remove the update-okta-oidc-tck?

File: composer.json [hunk 9]
```diff
@@ -27,28 +27,27 @@
...
+            "composer setup-env",
             "if [[ -f testenv.resource-server ]]; then source testenv.resource-server; fi && npm run test:resource-server"
         ],
         "sample:custom-login": "php -S 127.0.0.1:8080 -t ./custom-login ./custom-login/index.php",
         "sample:okta-hosted-login": "php -S 127.0.0.1:8080 -t ./okta-hosted-login ./okta-hosted-login/index.php",
         "sample:resource-server": "php -S 127.0.0.1:8000 -t ./resource-server ./resource-server/index.php",
-        "get-okta-oidc-tck": "if [[ `echo npm -v` ]]; then npm run get-oidc-tck; fi",
-        "update-okta-oidc-tck": "if [[ `echo npm -v` ]]; then npm run update-oidc-tck; fi",
```
Comment (bretterer) [inline composer.json]: Why did you remove the update-okta-oidc-tck?

File: scripts/setup-env.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  const envFile = path.join(directory, '.env');
+  const data = fs.readFileSync(envFile, 'utf8');
+
+  let clientId = process.env.CLIENT_ID;
+  // For resource server, we need to set client_id of SPA in .env
+  if (directory.includes('resource-server')) {
+    if (!process.env.SPA_CLIENT_ID) {
+      console.error('[ERROR] Please set the SPA_CLIENT_ID for resource-server tests');
```
Comment (bretterer) [inline scripts/setup-env.js:37]: Again, why can we not use the same application?  This worked fine for my testing?

File: scripts/setup-env.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+  if (process.platform === 'win32') {
+    execSync(`copy ${path.join(oktaHostedLoginDir, '.env.dist')} ${path.join(oktaHostedLoginDir, '.env')}`);
+    execSync(`copy ${path.join(customLoginDir, '.env.dist')} ${path.join(customLoginDir, '.env')}`);
+    execSync(`copy ${path.join(resourceServerDir, '.env.dist')} ${path.join(resourceServerDir, '.env')}`);
+  } else {
+    execSync(`cp ${path.join(oktaHostedLoginDir, '.env.dist')} ${path.join(oktaHostedLoginDir, '.env')}`);
+    execSync(`cp ${path.join(customLoginDir, '.env.dist')} ${path.join(customLoginDir, '.env')}`);
+    execSync(`cp ${path.join(resourceServerDir, '.env.dist')} ${path.join(resourceServerDir, '.env')}`);
+  }
```
Comment (bretterer) [inline scripts/setup-env.js]: I don't really want to automatically do this for users because it should be known that setting these vars is requires.  it is common practice in PHP to have a step of copying `.env.dist` over to `.env` in your setup guide.

File: CONTRIBUTING.md (hunk 1 above)
Comment (vijetmahabaleshwar-okta) [reply CONTRIBUTING.md:96]: For resource server, we need an SPA access token. If I use the same client_id in `.env` of resource-server, the tests fail.

File: CONTRIBUTING.md (hunk 4 above)
Comment (vijetmahabaleshwar-okta) [reply CONTRIBUTING.md]: Actually it's already being done in the `composer.json` scripts. Let me update the instructions.

File: composer.json (hunk 5 above)
Comment (vijetmahabaleshwar-okta) [reply composer.json:30]: The script `setup-env.js` updates the TCK when it's run

File: scripts/setup-env.js (hunk 10 above)
Comment (vijetmahabaleshwar-okta) [reply scripts/setup-env.js:37]: Try to copy `.env` that you used for `custom-login` into `resource-server`. I don't think it'll work

File: scripts/setup-env.js (hunk 11 above)
Comment (vijetmahabaleshwar-okta) [reply scripts/setup-env.js]: This script will be run only when running tests on CR. If you look at the READMEs, you wont see any references to this. 
Moreover the user can always manually copy the `.env`

Comment (bretterer) [reply scripts/setup-env.js]: when the script is run, will this over-write the existing .env file if the user manually provided one?  I don't see any checks here to make sure it does not exist first

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -71,6 +71,69 @@ $ git rebase master
...
+- `CLIENT_SECRET`
+- `SPA_CLIENT_ID`
+- `USER_NAME`
+- `PASSWORD`
+
+As an alternative you can provide the environment variables in a file named `testenv` in the root folder.
+
+For example:
+
+```
```
Comment (vijetmahabaleshwar-okta) [inline CONTRIBUTING.md:113]: Verified that providing `testenv` file works both in windows and linux.
Instead of having 2 env files earlier (`testenv`, `testenv.resource-server`), I suggest we have a single `testenv` that contains all env vars needed for testing login-server and resource-server.
Let me know what you think @bretterer

File: composer.json [hunk 13]
```diff
@@ -13,42 +13,42 @@
...
         ],
         "test:okta-hosted-login": [
-            "composer update-okta-oidc-tck",
-            "if [[ -f testenv ]]; then source testenv; fi && npm run test:okta-hosted-login"
+            "composer setup-env",
+            "npm run test:okta-hosted-login"
         ],
         "test:resource-server": [
-            "composer update-okta-oidc-tck",
-            "if [[ -f testenv.resource-server ]]; then source testenv.resource-server; fi && npm run test:resource-server"
```
Comment (vijetmahabaleshwar-okta) [inline composer.json:45]: moved this logic out to `setup-env.js` script.

File: composer.json [hunk 14]
```diff
@@ -13,42 +13,42 @@
     },
     "scripts": {
         "post-install-cmd": [
-            "composer get-okta-oidc-tck",
-            "if [[ `echo npm -v` ]]; then npm install; fi",
```
Comment (vijetmahabaleshwar-okta) [inline composer.json:17]: Removed this check for npm installation as it was not working on windows.
If the user doesn't have npm installed, the script will anyway fail.
Will probably need to update README or CONTRIBUTING.md to indicate that npm is a dependency

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -71,6 +71,69 @@ $ git rebase master
...
+ISSUER=https://dev-12345.oktapreview.com/oauth2/default
+CLIENT_ID=webclient123
+CLIENT_SECRET=websecret123
+SPA_CLIENT_ID=spaclient123
+USERNAME=myuser@example.com
+PASSWORD=mypassword
```
Comment (jmelberg-okta) [inline CONTRIBUTING.md:119]: Lets make sure the values we supply here match the above `export` commands. We should be throwing errors immediately if `{yourOktaDomain}` is the set `ISSUER`, while `dev-12345` will not throw.

File: scripts/setup-env.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  fs.writeFileSync(envFile, result, 'utf8');
+}
+
+function updateAllConfigs() {
+  const oktaHostedLoginDir = path.join(__dirname, '..', 'okta-hosted-login');
+  const customLoginDir = path.join(__dirname, '..', 'custom-login');
+  const resourceServerDir = path.join(__dirname, '..', 'resource-server');
+
+  if (process.platform === 'win32') {
+    execSync(`copy ${path.join(oktaHostedLoginDir, '.env.dist')} ${path.join(oktaHostedLoginDir, '.env')}`);
```
Comment (jmelberg-okta) [inline scripts/setup-env.js]: Doesn't need to be picked up with this PR - but we can simplify this quite a bit:

```javascript
const cmd = process.platform === 'win32' ? 'copy' : 'cp';

execSync(`${cmd} ${path.join(oktaHostedLoginDir, '.env.dist')} ${path.join(oktaHostedLoginDir, '.env')}`);
...

```

**Review by jmelberg-okta — APPROVED** [review]
Couple of suggestions. Once addressed 🚢

File: CONTRIBUTING.md (hunk 15 above)
Comment (vijetmahabaleshwar-okta) [reply CONTRIBUTING.md:119]: This is not a file that will be checked in. The user has to create a `testenv` file and provide the values manually. I don't think there is a chance that users will provide `{yourOktaDomain}` in it.

**Comment by bretterer** [comment]
Thank you for the work here!
